--- a/mt7915/main.c
+++ b/mt7915/main.c
@@ -1936,7 +1936,7 @@ mt7915_net_fill_forward_path(struct ieee
 	path->dev = ctx->dev;
 	path->mtk_wdma.wdma_idx = wed->wdma_idx;
 	path->mtk_wdma.bss = mvif->mt76.idx;
-	path->mtk_wdma.wcid = is_mt7915(&dev->mt76) ? 0xff : 0x3ff;
+	path->mtk_wdma.wcid = (is_mt7915(&dev->mt76) || test_bit(MT_WCID_FLAG_4ADDR, &msta->wcid.flags)) ? msta->wcid.idx : 0x3ff;
 	path->mtk_wdma.queue = phy != &dev->phy;
 
 	ctx->dev = NULL;
